### PR TITLE
Unserialize whitelist

### PR DESCRIPTION
**Pull request by MatthiasKuehneEllerhold**
See https://github.com/zendframework/zend-serializer/issues/36

### CODE REVIEW

File: src/Adapter/PhpSerialize.php [hunk 1]
```diff
@@ -35,6 +42,36 @@ public function __construct($options = null)
         parent::__construct($options);
     }
 
+    /**
+     * Set options
+     *
+     * @param  array|\Traversable|PhpSerializeOptions $options
```
Comment (marc-mabe) [inline src/Adapter/PhpSerialize.php]: please import `Traversable`

File: src/Adapter/PhpSerializeOptions.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (marc-mabe) [inline src/Adapter/PhpSerializeOptions.php]: 2015?

File: test/Adapter/PhpSerializeTest.php [hunk 3]
```diff
@@ -165,4 +165,49 @@ public function testUnserializingInvalidStringRaisesException($string, $expected
         $this->expectExceptionMessage($expected);
         $this->adapter->unserialize($string);
     }
+
+    public function testUnserializeNoWhitelistedClasses()
+    {
+        $value = 'O:8:"stdClass":0:{}';
+
+        $this->adapter->getOptions()->setUnserializeClassWhitelist(false);
```
Comment (marc-mabe) [inline test/Adapter/PhpSerializeTest.php]: This will fail on PHP < 7

File: test/Adapter/PhpSerializeTest.php [hunk 4]
```diff
@@ -165,4 +165,49 @@ public function testUnserializingInvalidStringRaisesException($string, $expected
...
+            // Pre PHP 7.0 the whitelist should have no effect
+            $this->assertInstanceOf(\stdClass::class, $data);
+        }
+    }
+
+    public function testUnserializeClassNotAllowed()
+    {
+        $value = 'O:8:"stdClass":0:{}';
+
+        $this->adapter->getOptions()->setUnserializeClassWhitelist([\My\Dummy::class]);
```
Comment (marc-mabe) [inline test/Adapter/PhpSerializeTest.php]: This will fail on PHP < 7

File: test/Adapter/PhpSerializeTest.php [hunk 5]
```diff
@@ -165,4 +165,49 @@ public function testUnserializingInvalidStringRaisesException($string, $expected
...
+            // Pre PHP 7.0 the whitelist should have no effect
+            $this->assertInstanceOf(\stdClass::class, $data);
+        }
+    }
+
+    public function testUnserializeClassAllowed()
+    {
+        $value = 'O:8:"stdClass":0:{}';
+
+        $this->adapter->getOptions()->setUnserializeClassWhitelist([\stdClass::class]);
```
Comment (marc-mabe) [inline test/Adapter/PhpSerializeTest.php]: This will fail on PHP < 7

File: src/Adapter/PhpSerializeOptions.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+     * Array of class names that are allowed to be unserialized
+     * or true if all classes should be allowed (behavior of pre PHP 7.0)
+     * or false if no classes should be allowed
+     *
+     * @var array|bool
+     */
+    protected $unserializeClassWhitelist = true;
+
+    /**
+     * @param  array|bool $unserializeClassWhitelist
```
Comment (marc-mabe) [inline src/Adapter/PhpSerializeOptions.php]: `@param  string[]|bool $unserializeClassWhitelist`

File: src/Adapter/PhpSerializeOptions.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                'Class whitelist for unserialize() is only available on PHP 7.0 or higher.'
+            );
+        }
+
+        $this->unserializeClassWhitelist = $unserializeClassWhitelist;
+        return $this;
+    }
+
+    /**
+     * @return array|bool
```
Comment (marc-mabe) [inline src/Adapter/PhpSerializeOptions.php]: `@return string[]|bool`

**Review by marc-mabe — COMMENTED** [review]
Else - from my perspective it looks quite good 👍

**Comment by weierophinney** [comment]
Thanks, @MatthiasKuehneEllerhold!

**Pull request merged**